### DISCUSSION TITLE
Ignore the updated event when an Eloquent model is restored

### DISCUSSION THREAD

**Discussion by medilies** [laravel/framework#43473] [Q&A]
I am building a [package](https://github.com/elaborate-code/laravel-eloquent-logs) that logs changes that happen on eloquent and I'm trying to figure out what to check to ignore the `updated` event when `restored`.

```php
trait HasLogs
{
    public static function bootHasLogs(): void
    {
        // created ...
        self::updated(callback: fn ($model) => self::log($model, 'updated'));

        self::deleted(callback: function ($model) {
            // Ignore 'deleted' event when 'softDeleted' or 'forceDeleted'
            if (in_array('Illuminate\Database\Eloquent\SoftDeletes', (class_uses(self::class)))) {
                return;
            }
            self::log($model, 'deleted');
        });

        if (in_array('Illuminate\Database\Eloquent\SoftDeletes', (class_uses(self::class)))) {
            // softDeleted ...
            self::restored(callback: fn ($model) => self::log($model, 'restored'));
            // forceDeleted ...
        }
    }
    // ...
}
```

Here is the actions order if that helps:

1. `Model::restore()` on the `SoftDeletes` trait.
2. ---- `restoring` is fired.
3. ---- `Model::save()`
4. -------- `saving` is fired.
5. -------- `Model::performUpdate()`
6. ------------ `updating` is fired.
7. ------------ `Builder::update()`
8. ------------ **`updated` is fired**.
9. -------- `Model::finishSave()`
10. ------------ `saved` is fired.
11. ---- `restored` is fired.

Original post: https://stackoverflow.com/questions/73158372/ignore-the-updated-event-when-an-eloquent-model-is-restored

**Comment by henzeb** [comment]
What does your package do that Spatie's isn't? https://spatie.be/docs/laravel-activitylog/v4/advanced-usage/logging-model-events

**Reply by medilies** [reply]
There is a huge difference when it comes to the sizes and complexity of the packages. And here is what I noticed.

## Similarities

### Registering event listeners

Both hook event listeners through traits with closures.

### Recording the actor

Both log which user made the change but in very different ways.

### Accessing the logs with subjects

Both offer a model that gives access to all activities with their subjects.

### Log has descriptions

The activity description/action_name is one word (created, updated, deleted ...).

But you can edit it on Spatie's package.

## Differences

### SoftDeletes trait events

eloquent-logs is explicit with `SoftDeletes` trait events, so more accuracy (soft deleted, force deleted, and restored).

### Archiving

activitylog archives the model states by saving the changed attributes.
I don't want to cover that feature to keep the package lightweight.

### Configuration

- activitylog is configurable allowing you to set:
  - What activity to log.
  - What attributes to watch.

Which I plan on adding.
